src/main/java/com/google/cloud/genomics/utils/grpc/GenomicsStreamIterator.java

```diff
@@ -0,0 +1,240 @@
+    while (true) {
+      try {
+        return createIteratorFromStub(request);
+      } catch (Exception e) {
+        if (shouldRetryNow()) {
+          LOG.log(Level.WARNING, "Retrying after failure to create iterator", e);
+        } else {
+          LOG.log(Level.WARNING, "All retries to create iterator consumed, re-throwing exception",
+              e);
+          throw e;
+        }
+      }
+    }
+  }
+
+  private boolean shouldRetryNow() {
+    long backOffMillis;
+    try {
+      backOffMillis = backoff.nextBackOffMillis();
+    } catch (IOException e1) {
+      // Something strange happened, just give up.
+      backOffMillis = BackOff.STOP;
+    }
+
+    if (backOffMillis == BackOff.STOP) {
+      backoff.reset();
+      return false;
+    }
+
+    try {
+      Thread.sleep(backOffMillis);
+    } catch (InterruptedException e) {
+      LOG.log(Level.WARNING, "Backoff sleep interrupted", e);
+    }
+    return true;
+  }
+
+  @Override
+  protected Iterator<B> delegate() {
+    return delegate;
+  }
+
+  /**
+   * @see java.util.Iterator#hasNext()
+   */
+  public boolean hasNext() {
+    boolean hasNext;
+    while (true) {
+      try {
+        hasNext = delegate.hasNext();
+        break;
+      } catch (Exception e) {
+        if (shouldRetryNow()) {
+          LOG.log(Level.WARNING, "Retrying after failing to get next item from stream: ", e);
+          setStreamStateForRetry();
+        } else {
+          LOG.log(Level.WARNING, "All retries to get next item from stream consumed, throwing: ", e);
+          genomicsChannel.shutdownNow();
+          throw e;
+        }
+      }
+    }
+    if (!hasNext) {
+      genomicsChannel.shutdownNow();
+    }
+    return hasNext;
+  }
+
+  private void setStreamStateForRetry() {
+    if (null == lastSuccessfulDataItem) {
+      // We have never returned any data. No need to set up state needed to filter previously
+      // returned results.
+      return;
+    }
+
+    if (getRequestStart(originalRequest) < getDataItemStart(lastSuccessfulDataItem)) {
+      // Create a new iterator at the revised start position.
+      delegate = createIterator(getRevisedRequest(getDataItemStart(lastSuccessfulDataItem)));
+    } else {
+      // The point at which the retry occurred was still within data overlapping the start of our
+      // original request but not beyond it yet.
+      delegate = createIterator(originalRequest);
+    }
+
+    // RETRY STATE: Enable the filtering of repeated data in next().
+    idSentinel = getDataItemId(lastSuccessfulDataItem);
+  }
+
+  /**
+   * @see java.util.Iterator#next()
+   */
+  public B next() {
+    B response = delegate.next();
+    // TODO: Its more clean conceptually to do the same thing for all responses, but this could be a
+    // place where we're wasting a lot of time rebuilding response objects when nothing has actually
+    // changed.
+    return setDataList(response, enforceShardPredicate(removeRepeatedData(getDataList(response))));
```

Comment:
thanks for the ping on this

This is equivalent to [how we've been doing it all along](https://github.com/googlegenomics/utils-java/blob/master/src/main/java/com/google/cloud/genomics/utils/Paginator.java#L401) (possibly returning an empty list).

It would be a good idea to call this out explicitly in an integration test.